platform/jvm/capture/src/main/kotlin/io/bitdrift/capture/attributes/NetworkAttributes.kt

```diff
@@ -133,29 +105,38 @@ internal class NetworkAttributes(
     }
 
     private fun updateNetworkType(networkCapabilities: NetworkCapabilities?) {
-        currentNetworkType.set(
+        val type =
             networkCapabilities?.run {
                 when {
                     hasTransport(TRANSPORT_WIFI) -> "wlan"
                     hasTransport(TRANSPORT_CELLULAR) -> "wwan"
                     hasTransport(TRANSPORT_ETHERNET) -> "ethernet"
                     else -> "other"
                 }
-            } ?: "unknown",
-        )
+            } ?: "unknown"
+
+        updateFields(networkType = type)
     }
 
     private fun updateTelephonyAttributes() {
-        val newCarrier = telephonyManager.networkOperatorName ?: "unknown"
-        val newRadioType = permissiveOperation({ radioType() }, READ_PHONE_STATE)
+        val carrier = telephonyManager.networkOperatorName ?: "unknown"
+        val radioType = permissiveOperation({ radioType() }, READ_PHONE_STATE)
 
-        if (currentCarrier.get() != newCarrier) {
-            currentCarrier.set(newCarrier)
-        }
+        updateFields(carrier = carrier, radioType = radioType)
+    }
 
-        if (currentRadioType.get() != newRadioType) {
-            currentRadioType.set(newRadioType)
-        }
+    private fun updateFields(
+        carrier: String? = null,
+        networkType: String? = null,
+        radioType: String? = null,
+    ) {
+        val updated =
+            currentFields.toMutableMap().apply {
+                carrier?.let { this[KEY_CARRIER] = it }
+                networkType?.let { this[KEY_NETWORK_TYPE] = it }
+                radioType?.let { this[KEY_RADIO_TYPE] = it }
+            }
+        currentFields = updated.toMap()
```

Comment:
You should double check, but I believe both `toMutableMap()` and `toMap()` always create a copy. If I'm right, given that `NetworkAttributes` is internal, and assuming the consuming code doesn't try to make any change to the map returned by `fun invoke(): Fields`, you could remove the `toMap()` here, save yourself a copy. That map will never be updated again